src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/TdsParser.cs

```diff
@@ -10231,6 +9797,493 @@ internal Task TdsExecuteRPC(SqlCommand cmd, IList<_SqlRPC> rpcArray, int timeout
+                }
+            }
+            else if (mt.IsCharType && mt.SqlDbType != SqlDbTypeExtensions.Json)
+            {
+                // if it is not supplied, simply write out our default collation, otherwise, write out the one attached to the parameter
+                SqlCollation outCollation = (param.Collation != null) ? param.Collation : _defaultCollation;
+                Debug.Assert(_defaultCollation != null, "_defaultCollation is null!");
+
+                WriteUnsignedInt(outCollation._info, stateObj);
+                stateObj.WriteByte(outCollation._sortId);
+            }
+
+            if (0 == codePageByteSize)
+                WriteParameterVarLen(mt, actualSize, isNull, stateObj, isDataFeed);
+            else
+                WriteParameterVarLen(mt, codePageByteSize, isNull, stateObj, isDataFeed);
+
+            Task writeParamTask = null;
+            // write the value now
+            if (!isNull)
+            {
+                if (isSqlVal)
+                {
+                    writeParamTask = WriteSqlValue(value, mt, actualSize, codePageByteSize, param.Offset, stateObj);
+                }
+                else
+                {
+                    // for codePageEncoded types, WriteValue simply expects the number of characters
+                    // For plp types, we also need the encoded byte size
+                    writeParamTask = WriteValue(value, mt, isParameterEncrypted ? (byte)0 : param.GetActualScale(), actualSize, codePageByteSize, isParameterEncrypted ? 0 : param.Offset, stateObj, isParameterEncrypted ? 0 : param.Size, isDataFeed);
+                }
+            }
+
+            // Send encryption metadata for encrypted parameters.
+            if (isParameterEncrypted)
+            {
+                writeParamTask = WriteEncryptionMetadata(writeParamTask, encryptedParameterInfoToWrite, stateObj);
+            }
+
+            return writeParamTask;
+        }
+
+        // This is in its own method to avoid always allocating the lambda in TDSExecuteRPCParameter
+        private void TDSExecuteRPCParameterSetupWriteCompletion(SqlCommand cmd, IList<_SqlRPC> rpcArray, int timeout, bool inSchema, SqlNotificationRequest notificationRequest, TdsParserStateObject stateObj, bool isCommandProc, bool sync, TaskCompletionSource<object> completion, int startRpc, int startParam, Task writeParamTask)
+        {
+            AsyncHelper.ContinueTask(
+                writeParamTask,
+                completion,
+                () => TdsExecuteRPC(
+                    cmd,
+                    rpcArray,
+                    timeout,
+                    inSchema,
+                    notificationRequest,
+                    stateObj,
+                    isCommandProc,
+                    sync,
+                    completion,
+                    startRpc,
+                    startParam
+                ),
+                onFailure: exc => TdsExecuteRPC_OnFailure(exc, stateObj)
+            );
```

Comment:
This corresponds to lines 10138-10146, and it swaps `AsyncHelper.ContinueTaskWithState` for `AsyncHelper.ContinueTask`. The passing of state only seems to have an effect on the `onFailure` lambda, but we also no longer pass `connectionToDoom`.

Reply:
Are you saying I should be replacing this code with what's in 10138-10146 in main? Or just noting what the difference is for others. 
Doing the rebase on this one was a bit tricky (since I don't trust myself touching TdsParser), but I generally just took whatever the newest code was. In this case, @MichelZ's version was newer (the previous was written by @Wraith2 in 2021), so I took that.

Reply:
If I remember correctly the connection to doom is present in netfx but not netcore. The difference between ContinueTask and ContinueTaskWithState should only be that the second avoids a closure on the common path, closures on exception paths are irrelevant since it's going to allocate exception, errors, stack trace etc.

Reply:
I agree that the closure is probably a drop in the ocean (allocation-wise) - but I think we should replace the code with L10138-10146. It's not worth removing the closure's state management, and the exception handling which comes with connectionToDoom would be a regression on netfx. The connectionToDoom parameter is present in both targets now, it looks like that was ported over pretty recently.

Looking more broadly at TdsParser, it's surprisingly similar to SqlConnection - there's a fairly large diff, but it's mostly comprised of non-functional changes. TdsExecuteRPC makes up about 20% of that diff.

Reply:
The core, TDSParser, SqlConnectionInternal are fairly heavily conserved between the two codebases. Most people are rightfully afraid to touch them unless required.

Reply:
Ok, I replaced it with L10138-10146. Hopefully that's what the agreement seems to be 😬